text/000-structural-and-deref.md

```diff
@@ -0,0 +1,410 @@
+- Start Date: 2018-10-05
+- RFC PR: (leave this empty)
+- Tracking Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Change `#[wasm_bindgen]` to use `structural` by default, and add a new
+attribute `final` for an opt-in to today's behavior. Once implemented then use
+`Deref` to model the class inheritance hierarchy in `web-sys` and `js-sys` to
+enable ergonomic usage of superclass methods of web types.
+
+# Motivation
+[motivation]: #motivation
+
+The initial motivation for this is outlined [RFC 3], namely that the `web-sys`
+crate provides bindings for many APIs found on the web but accessing the
+functionality of parent classes is quite cumbersome.
+
+The web makes extensive use of class inheritance hierarchies, and in `web-sys`
+right now each class gets its own `struct` type with inherent methods. These
+types implement `AsRef` between one another for subclass relationships, but it's
+quite unergonomic to actually reference the functionality! For example:
+
+```rust
+let x: &Element = ...;
+let y: &Node = x.as_ref();
+y.append_child(...);
+```
+
+or...
+
+```rust
+let x: &Element = ...;
+<Element as AsRef<Node>>::as_ref(x)
+    .append_child(...);
+```
+
+It's be much nicer if we could support this in a more first-class fashion and
+make it more ergonomic!
+
+[RFC 3]: https://github.com/rustwasm/rfcs/pull/3
+
+> **Note**: While this RFC has the same motivation as [RFC 3] it's proposing an
+> alternative solution, specifically enabled by switching by `structural` by
+> default, which is discussed in [RFC 3] but is hopefully formally outlined
+> here.
+
+# Detailed Explanation
+[detailed-explanation]: #detailed-explanation
+
+This RFC proposes using the built-in `Deref` trait to model the class hierarchy
+found on the web in `web-sys`. This also proposes changes to `#[wasm_bindgen]`
+to make using `Deref` feasible for binding arbitrary JS apis (such as those on
+NPM) with `Deref` as well.
+
+For example, `web-sys` will contain:
+
+```rust
+impl Deref for Element {
+    type Target = Node;
+
+    fn deref(&self) -> &Node { /* ... */ }
+}
+```
+
+allowing us to write our example above as:
+
+```rust
+let x: &Element = ...;
+x.append_child(...); // implicit deref to `Node`!
+```
+
+All JS types in `web-sys` and in general have at most one superclass. Currently,
+however, the `#[wasm_bindgen]` attribute allows specifying multiple `extends`
+attributes to indicate superclasses:
+
+```rust
+#[wasm_bindgen]
+extern {
+    #[wasm_bindgen(extends = Node, extends = Object)]
+    type Element;
+
+    // ...
+}
+```
+
+The `web-sys` API generator currently lists an `extends` for all superclasses,
+transitively. This is then used in the code generator to generate `AsRef`
+implementatiosn for `Element`.
+
+The code generation of `#[wasm_bindgen]` will be updated with the following
```

Comment:
This seems like a footgun. Perhaps it could be split into two separate attributes, like this?

```rust
#[wasm_bindgen(parent = Node, extends = Object)]
```

So that way it's crystal clear what's going on, and it's harder to make mistakes (and if somebody does make a mistake, such as using `extends` without `parent`, it can print a nice error message).

Reply:
Yes I'd be fine adding that in the future, but for now allowing multiple `extends` while still implementing `Deref` is a backwards-compatibility requirement.

Reply:
I also think long term we don't want a new attribute, but rather only accepting one `extends` attribute.

Reply:
@alexcrichton Is it a backwards compatibility requirement? The RFC itself says that web-sys will have to be changed so that it always puts the immediate parent first, so this requirement requires code changes. The same is true for non-web-sys bindings as well.

So I don't see how this enables backwards compatibility. In fact, it may even be a huge footgun: some existing code might put the wrong class first, but this doesn't give a warning or error, so they don't realize it.

If we instead required a new `parent` attribute, we would give a nice error message if it's missing, and this lets people know that they need to change their code so that the immediate parent is correct.

Reply:
Yes this would otherwise be a breaking change for `wasm-bindgen`, which we're not doing at this time. `web-sys` can be updated easily, other crates cannot.

Reply:
Somehow that still feels wrong. If there exists code where the first `extends` isn't the immediate parent, then they will now get an automatic `Deref` impl which is wrong (or at least unexpected).

We could instead tie `Deref` to the `parent` attribute, so that if code wants the `Deref` impl they have to change to use `parent` (existing code without `parent` will keep the old behavior and not have a `Deref` impl).